C error (when dumping Option pointer array): invalid operand types for binary operation

### Describe the bug

I get a C error when dumping an array of option of pointers

### Expected Behavior

I expect the code to work

### Current Behavior

v -cg output:
```
C:/Users/Usuario/AppData/Local/Temp/v_0/07.5427593099956853890.tmp.c:501: warning: WINVER redefined
C:/Users/Usuario/AppData/Local/Temp/v_0/07.5427593099956853890.tmp.c:2218: error: invalid operand types for binary operation
C:/Users/Usuario/AppData/Local/Temp/v_0/07.5427593099956853890.tmp.c:2218: error: invalid aggregate type for register load
builder error:
==================
C error. This should never happen.
```

### Reproduction Steps

run the following code:
```
fn main(){
	mut a := []?&int{}
	dump(a)
}
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.4 a12e82a

### Environment details (OS name and version, etc.)

V full version: V 0.3.4 34163ae.a12e82a
OS: windows, Microsoft Windows 10 Pro v19045 64-bit
Processor: 8 cpus, 64bit, little endian,

getwd: C:\Users\Usuario\Desktop\Universidad\mag\v_test\07
vexe: C:\Program Files\v\v.exe
vexe mtime: 2023-06-22 14:13:17

vroot: contains spaces, value: C:\Program Files\v
VMODULES: OK, value: C:\Users\Usuario\.vmodules
VTMP: OK, value: C:\Users\Usuario\AppData\Local\Temp\v_0

Git version: git version 2.40.1.windows.1
Git vroot status: weekly.2023.25-15-ga12e82aa
.git/config present: true

CC version: cc (Rev6, Built by MSYS2 project) 13.1.0
thirdparty/tcc status: thirdparty-windows-amd64 e90c2620

--- a/vlib/v/checker/infix.v
+++ b/vlib/v/checker/infix.v
@@ -522,6 +522,9 @@ fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 				} else if c.check_types(unwrapped_right_type, c.unwrap_generic(left_type)) {
 					return ast.void_type
 				}
+				if left_value_type.has_flag(.option) && right_type == ast.none_type {
+					return ast.void_type
+				}
 				c.error('cannot append `${right_sym.name}` to `${left_sym.name}`', right_pos)
 				return ast.void_type
 			} else {

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -578,10 +578,14 @@ fn (mut g Gen) gen_str_for_array(info ast.Array, styp string, str_fn_name string
 				deref, deref_label := deref_kind(str_method_expects_ptr, is_elem_ptr,
 					typ)
 				g.auto_str_funcs.writeln('\t\tstring x = _SLIT("nil");')
-				g.auto_str_funcs.writeln('\t\tif (it != 0) {')
+				if !typ.has_flag(.option) {
+					g.auto_str_funcs.writeln('\t\tif (it != 0) {')
+				}
 				g.auto_str_funcs.writeln('\t\t\tstrings__Builder_write_string(&sb, _SLIT("${deref_label}"));')
 				g.auto_str_funcs.writeln('\t\t\tx = ${elem_str_fn_name}(${deref}it);')
-				g.auto_str_funcs.writeln('\t\t}')
+				if !typ.has_flag(.option) {
+					g.auto_str_funcs.writeln('\t\t}')
+				}
 			} else {
 				g.auto_str_funcs.writeln('\t\tstring x = indent_${elem_str_fn_name}(it, indent_count);')
 			}
@@ -610,10 +614,14 @@ fn (mut g Gen) gen_str_for_array(info ast.Array, styp string, str_fn_name string
 			deref, deref_label := deref_kind(str_method_expects_ptr, is_elem_ptr, typ)
 			if is_elem_ptr {
 				g.auto_str_funcs.writeln('\t\tstring x = _SLIT("nil");')
-				g.auto_str_funcs.writeln('\t\tif (it != 0) {')
+				if !typ.has_flag(.option) {
+					g.auto_str_funcs.writeln('\t\tif (it != 0) {')
+				}
 				g.auto_str_funcs.writeln('\t\t\tstrings__Builder_write_string(&sb, _SLIT("${deref_label}"));')
 				g.auto_str_funcs.writeln('\t\t\tx = ${elem_str_fn_name}(${deref}it);')
-				g.auto_str_funcs.writeln('\t\t}')
+				if !typ.has_flag(.option) {
+					g.auto_str_funcs.writeln('\t\t}')
+				}
 			} else {
 				g.auto_str_funcs.writeln('\t\tstrings__Builder_write_string(&sb, _SLIT("${deref_label}"));')
 				g.auto_str_funcs.writeln('\t\tstring x = ${elem_str_fn_name}(${deref}it);')

--- a/vlib/v/gen/c/infix.v
+++ b/vlib/v/gen/c/infix.v
@@ -990,7 +990,7 @@ fn (mut g Gen) infix_expr_and_or_op(node ast.InfixExpr) {
 }
 
 fn (mut g Gen) gen_is_none_check(node ast.InfixExpr) {
-	if node.left in [ast.Ident, ast.SelectorExpr] {
+	if node.left in [ast.Ident, ast.SelectorExpr, ast.IndexExpr] {
 		old_inside_opt_or_res := g.inside_opt_or_res
 		g.inside_opt_or_res = true
 		g.expr(node.left)

--- a/vlib/v/tests/option_assign_none_test.v
+++ b/vlib/v/tests/option_assign_none_test.v
@@ -1,0 +1,11 @@
+fn test_main() {
+	mut a := []?&int{}
+	a << ?&int(none)
+	a << none
+	dump(a[0])
+	dump(a[1])
+	println(a)
+
+	assert a[0] == none
+	assert a[1] == none
+}

